packages/mds-agency/handlers/create-event.ts

```diff
@@ -0,0 +1,168 @@
+ */
+const logEventWritePerformance = (event: VehicleEvent, logThreshold = 100) => {
+  const { recorded } = event
+  const delta = now() - recorded
+
+  /* Tests shouldn't be slow */
+  /* istanbul ignore next */
+  if (delta > logThreshold) {
+    const { provider_id } = event
+    const name = providerName(provider_id || 'unknown')
+
+    logger.info(`${name} post event took ${delta} ms`)
+  }
+}
+
+const sendSuccess = (
+  req: AgencyApiSubmitVehicleEventRequest,
+  res: AgencyApiSubmitVehicleEventResponse,
+  event: VehicleEvent
+) => {
+  const { device_id, vehicle_state } = event
+
+  logEventWritePerformance(event)
+
+  return res.status(201).send({
+    device_id,
+    state: vehicle_state
+  })
+}
+
+/**
+ * Refreshes the cache if a device was previously registered, but was removed from the cache due to decommissioning
+ * @param device MDS Device
+ * @param event MDS VehicleEvent
+ */
+const refreshDeviceCache = async (device: Device, event: VehicleEvent) => {
+  try {
+    await cache.readDevice(event.device_id)
+  } catch (err) {
+    try {
+      await Promise.all([cache.writeDevice(device), stream.writeDevice(device)])
+      logger.info('Re-adding previously deregistered device to cache', err)
+    } catch (error) {
+      logger.warn(`Error writing to cache/stream ${error}`)
+    }
+  }
+}
+
+export const createEventHandler = async (
+  req: AgencyApiSubmitVehicleEventRequest,
+  res: AgencyApiSubmitVehicleEventResponse
+) => {
+  const { device_id } = req.params
+
+  const { provider_id } = res.locals
+
+  const recorded = now()
+
+  const unparsedEvent = {
+    ...req.body,
+    device_id,
+    provider_id,
+    recorded
+  }
+
+  try {
+    const event = (() => {
+      const parsedEvent = validateEventDomainModel(unparsedEvent)
+
+      const { telemetry, device_id } = parsedEvent
+
+      if (telemetry) {
+        const { timestamp: telemetry_timestamp } = telemetry
+
+        return { ...parsedEvent, telemetry_timestamp, telemetry: { ...telemetry, device_id } }
+      }
+
+      return parsedEvent
+    })()
+
+    // TODO switch to cache for speed?
+    const device = await db.readDevice(device_id, provider_id)
+
+    // Note: Even though the event has passed schema validation, we need to verify it's allowed for this mode of vehicle
+    const invalidStateOrEventTypes = eventValidForMode(device, event)
+    if (invalidStateOrEventTypes) {
+      return res.status(400).send(invalidStateOrEventTypes)
+    }
+
+    await refreshDeviceCache(device, event)
+
+    const { telemetry } = event
+    if (telemetry) {
+      await db.writeTelemetry(normalizeToArray(telemetry))
+    }
+
+    // database write is crucial; failures of cache/stream should be noted and repaired
+    const recorded_event = await db.writeEvent(event)
+
+    try {
+      await Promise.all([cache.writeEvent(recorded_event), stream.writeEvent(recorded_event)])
+
+      if (telemetry) {
+        telemetry.recorded = recorded
```

Comment:
Isn't the `recorded` value for telemetry going to be different from the `recorded` value in the DB? Why does the `recorded` field on telemetry and events need to match up in the cache?

Reply:
That's a good question, this can probably be cleaned up now that telemetry is required events. I imagine there was probably a slight delta before, yup.

Reply:
Fixed in cff2d96